protocol/src/lib.rs

```diff
@@ -615,6 +617,104 @@ pub enum Message {
+pub struct WriteHeader {
+    pub upstairs_id: Uuid,
+    pub session_id: Uuid,
+    pub job_id: JobId,
+    pub dependencies: Vec<JobId>,
+    pub blocks: Vec<WriteBlockMetadata>,
+}
+
+impl WriteHeader {
+    /// Destructures into a list of block-size writes
+    ///
+    /// # Panics
+    /// `buf.len()` must be an even multiple of `self.blocks.len()`, which is
+    /// assumed to be the block size.
+    pub fn into_writes(&self, mut buf: bytes::Bytes) -> Vec<Write> {
+        assert_eq!(buf.len() % self.blocks.len(), 0);
+        let block_size = buf.len() / self.blocks.len();
+        let mut out = Vec::with_capacity(self.blocks.len());
+        for b in &self.blocks {
+            let data = buf.split_to(block_size);
+            out.push(Write {
+                eid: b.eid,
+                offset: b.offset,
+                block_context: b.block_context,
+                data,
+            })
+        }
+        out
+    }
+}
+
+#[derive(Debug, PartialEq, Copy, Clone, Serialize, Deserialize)]
+pub struct WriteBlockMetadata {
+    pub eid: u64,
+    pub offset: Block,
+    pub block_context: BlockContext,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct ReadResponseHeader {
+    pub upstairs_id: Uuid,
+    pub session_id: Uuid,
+    pub job_id: JobId,
+    pub blocks: Result<Vec<ReadResponseBlockMetadata>, CrucibleError>,
+}
+
+impl ReadResponseHeader {
+    /// Destructures into a list of block-size read responses
+    ///
+    /// # Panics
+    /// `buf.len()` must be an even multiple of `self.blocks.len()`, which is
+    /// assumed to be the block size.
+    pub fn into_read_responses(
+        self,
+        mut buf: bytes::BytesMut,
+    ) -> Result<Vec<ReadResponse>, CrucibleError> {
+        let blocks = self.blocks?;
+        assert_eq!(buf.len() % blocks.len(), 0);
+        let block_size = buf.len() / blocks.len();
+        let mut out = Vec::with_capacity(blocks.len());
+        for b in blocks {
+            let data = buf.split_to(block_size);
+            out.push(ReadResponse {
+                eid: b.eid,
+                offset: b.offset,
+                block_contexts: b.block_contexts,
+                data,
+            })
+        }
+        Ok(out)
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct ReadResponseBlockMetadata {
+    pub eid: u64,
+    pub offset: Block,
+    pub block_contexts: Vec<BlockContext>,
+}
+
+impl ReadResponseBlockMetadata {
+    pub fn hashes(&self) -> Vec<u64> {
+        self.block_contexts.iter().map(|x| x.hash).collect()
+    }
+
+    pub fn first_hash(&self) -> Option<u64> {
+        self.block_contexts.first().map(|ctx| ctx.hash)
+    }
+
+    pub fn encryption_contexts(&self) -> Vec<Option<&EncryptionContext>> {
+        self.block_contexts
+            .iter()
+            .map(|x| x.encryption_context.as_ref())
+            .collect()
+    }
+}
+
 /*
  * If you just added or changed the Message enum above, you must also
```

Comment:
I'm not sure what I deleted here?

The lines I removed from the `enum Message` were things like
```
// Message::Write must contain the same fields in the same order as
// RawMessage::WriteUnwritten, which is used for zero-copy serialization.
```

which _should_ be removed, since we're no longer using `RawMessage` entirely.

Reply:
I mean the `If you just added or changed the Message enum above` comment, it should be right after the Message enum ends.

Reply:
Ah, I see – the PR adds some stuff in between that comment and its desired location.

Fixed in 111b9a07f0caeab469d520ef27f4aebb37d9e93d